waithelper.go

```diff
@@ -52,62 +48,70 @@ func waitForSingleObject(handle syscall.Handle, timeout uint32) (bool, error) {
+func processAsyncHcsResult(err error, resultp *uint16, callbackNumber uintptr, expectedNotification hcsNotification, timeout *time.Duration) error {
 	err = processHcsResult(err, resultp)
 	if err == ErrVmcomputeOperationPending {
-		if timeout != nil {
-			err = registerAndWaitForCallbackTimeout(object, expectedNotification, *timeout)
-		} else {
-			err = registerAndWaitForCallback(object, expectedNotification)
-		}
+		return waitForNotification(callbackNumber, expectedNotification, timeout)
 	}
 
 	return err
 }
 
-func registerAndWaitForCallbackTimeout(object callbackable, expectedNotification hcsNotification, timeout time.Duration) error {
-	callbackNumber, err := object.registerCallback(expectedNotification)
-	if err != nil {
-		return err
-	}
-	defer object.unregisterCallback(callbackNumber)
-
-	return waitForNotificationTimeout(callbackNumber, timeout)
-}
-
-func registerAndWaitForCallback(object callbackable, expectedNotification hcsNotification) error {
-	callbackNumber, err := object.registerCallback(expectedNotification)
-	if err != nil {
-		return err
-	}
-	defer object.unregisterCallback(callbackNumber)
-
-	return waitForNotification(callbackNumber)
-}
-
-func waitForNotificationTimeout(callbackNumber uintptr, timeout time.Duration) error {
+func waitForNotification(callbackNumber uintptr, expectedNotification hcsNotification, timeout *time.Duration) error {
 	callbackMapLock.RLock()
-	channel := callbackMap[callbackNumber].channel
+	channels := callbackMap[callbackNumber].channels
 	callbackMapLock.RUnlock()
 
-	timer := time.NewTimer(timeout)
-	defer timer.Stop()
-
-	select {
-	case err := <-channel:
-		return err
-	case <-timer.C:
-		return ErrTimeout
+	var expectedChannel notificationChannel
+
+	switch expectedNotification {
+	case hcsNotificationSystemExited:
+		expectedChannel = channels.containerExitedChannel
+	case hcsNotificationSystemCreateCompleted:
+		expectedChannel = channels.containerCreateCompletedChannel
+	case hcsNotificationSystemStartCompleted:
+		expectedChannel = channels.containerStartCompletedChannel
+	case hcsNotificationSystemPauseCompleted:
+		expectedChannel = channels.containerPauseCompletedChannel
+	case hcsNotificationSystemResumeCompleted:
+		expectedChannel = channels.containerResumeCompletedChannel
+	case hcsNotificationProcessExited:
+		expectedChannel = channels.processExitedChannel
+	case hcsNotificationServiceDisconnect:
+		expectedChannel = channels.serviceDisconnectChannel
+	default:
+		logrus.Errorf("unknown notification type in waitForNotification %x", expectedNotification)
 	}
-}
-
-func waitForNotification(callbackNumber uintptr) error {
-	callbackMapLock.RLock()
-	channel := callbackMap[callbackNumber].channel
-	callbackMapLock.RUnlock()
 
+	if timeout != nil {
+		timer := time.NewTimer(*timeout)
+		defer timer.Stop()
+
+		select {
+		case err := <-expectedChannel:
+			return err
+		case err := <-channels.containerExitedChannel:
+			if channels.containerExitedChannel == expectedChannel {
```

Comment:
In this case will it not be the same issue with channels.serviceDisconnectChannel?